File: core/x/javaapi/api/src/main/java/viaduct/java/api/types/BackwardConnectionArguments.java
package viaduct.java.api.types;

import viaduct.tenant.runtime.support.ConnectionArgumentsSupport;

/**
 * Arguments for backward pagination through a connection.
 *
 * <p>Java equivalent of Kotlin's {@code viaduct.api.types.BackwardConnectionArguments}. The math
 * and validation are delegated to the language-neutral {@link ConnectionArgumentsSupport} shared
 * with the Kotlin Tenant API.
 *
 * @see ForwardConnectionArguments
 */
public interface BackwardConnectionArguments extends ConnectionArguments {
  /** Maximum number of items to return from the end; null for the default page size. */
  Integer getLast();

  /** Cursor to start fetching items before (exclusive); null to start from the end. */
  String getBefore();

  /**
   * Returns true if backward pagination needs the total count to resolve the starting offset.
   *
   * <p>This happens whenever {@link #getBefore()} is absent, including the default
   * backward-pagination case where {@link #getLast()} is also absent and the default page size is
   * used. Use {@link #toOffsetLimit(int, int)} instead of {@link #toOffsetLimit(int)}.
   */
  @Override
  default boolean requiresTotalCountForOffsetLimit() {
    return ConnectionArgumentsSupport.INSTANCE.backwardRequiresTotalCount(getBefore());
  }

  /**
   * Converts backward pagination arguments to offset/limit.
   *
   * <ul>
   *   <li>{@link #getLast()} determines the page size (defaults to {@code defaultPageSize}).
   *   <li>{@link #getBefore()} cursor encodes the index of the first item on the next page; the
   *       offset and limit are computed to return the last items ending just before that position,
   *       clamped so the window never extends before index 0.
   *   <li>If {@code before} is absent, a negative offset equal to {@code -pageSize} is returned,
   *       signalling {@link viaduct.java.api.internal.ConnectionBuilder#fromList} to resolve from
   *       the tail of the full list. Prefer {@link #toOffsetLimit(int, int)} with the total count
   *       when available.
   * </ul>
   */
  @Override
  default OffsetLimit toOffsetLimit(int defaultPageSize) {
    ConnectionArgumentsSupport.OffsetBounds b =
        ConnectionArgumentsSupport.INSTANCE.backwardOffsetLimit(
            getLast(), getBefore(), defaultPageSize);
    return new OffsetLimit(b.getOffset(), b.getLimit());
  }

  /**
   * Converts backward pagination arguments to offset/limit when the total count is known.
   *
   * <p>When {@link #getBefore()} is absent, this computes the offset as {@code max(0, totalCount -
   * last)} to return the last N items from the dataset, avoiding the negative-offset signal used by
   * {@link #toOffsetLimit(int)}. When {@code before} is present, {@code totalCount} is ignored and
   * this delegates to {@link #toOffsetLimit(int)}.
   *
   * @param totalCount total number of items in the full dataset
   * @param defaultPageSize default number of items when last not specified
   */
  @Override
  default OffsetLimit toOffsetLimit(int totalCount, int defaultPageSize) {
    ConnectionArgumentsSupport.OffsetBounds b =
        ConnectionArgumentsSupport.INSTANCE.backwardOffsetLimit(
            getLast(), getBefore(), totalCount, defaultPageSize);
    return new OffsetLimit(b.getOffset(), b.getLimit());
  }

  /**
   * Validates backward pagination arguments.
   *
   * @throws IllegalArgumentException if last is not positive or before cursor is invalid
   */
  @Override
  default void validate() {
    ConnectionArgumentsSupport.INSTANCE.validateBackward(getLast(), getBefore());
  }
}

File: core/tenant/api/src/main/kotlin/viaduct/api/types/BackwardConnectionArguments.kt
package viaduct.api.types

import viaduct.apiannotations.ExperimentalApi
import viaduct.tenant.runtime.support.ConnectionArgumentsSupport

/**
 * Arguments for backward pagination through a connection.
 *
 * The offset/limit math and validation are delegated to the language-neutral
 * [ConnectionArgumentsSupport] so the Kotlin and Java Tenant APIs stay in lockstep.
 *
 * @property last Maximum number of items to return from the end.
 * @property before Cursor to start fetching items before (exclusive).
 * @see ForwardConnectionArguments
 */
@ExperimentalApi
interface BackwardConnectionArguments : ConnectionArguments {
    val last: Int?
    val before: String?

    /**
     * Returns true if backward pagination needs the total count to resolve the starting offset.
     *
     * This happens whenever [before] is absent, including the default backward-pagination case
     * where [last] is also absent and the default page size is used. Use [toOffsetLimit(Int, Int)]
     * instead of [toOffsetLimit(Int)].
     */
    @ExperimentalApi
    override fun requiresTotalCountForOffsetLimit(): Boolean = ConnectionArgumentsSupport.backwardRequiresTotalCount(before)

    /**
     * Converts backward pagination arguments to offset/limit.
     *
     * - [last] determines the page size (defaults to [defaultPageSize]).
     * - [before] cursor encodes the index of the first item on the next page; the offset and limit
     *   are computed to return the last items ending just before that position, clamped so the
     *   window never extends before index 0.
     * - If [before] is absent, a negative offset equal to `-pageSize` is returned, signalling
     *   [viaduct.api.internal.ConnectionBuilder.fromList] to resolve from the tail of the full
     *   list. Prefer [toOffsetLimit(Int, Int)] with the total count when available.
     */
    @ExperimentalApi
    override fun toOffsetLimit(defaultPageSize: Int): OffsetLimit =
        ConnectionArgumentsSupport.backwardOffsetLimit(last, before, defaultPageSize).let {
            OffsetLimit(offset = it.offset, limit = it.limit)
        }

    /**
     * Converts backward pagination arguments to offset/limit when the total count is known.
     *
     * When [before] is absent, this computes the offset as `max(0, totalCount - last)` to return
     * the last N items from the dataset, avoiding the negative-offset signal used by
     * [toOffsetLimit(Int)]. When [before] is present, [totalCount] is ignored.
     *
     * @param totalCount Total number of items in the full dataset
     * @param defaultPageSize Default number of items when last not specified (default: 20)
     */
    @ExperimentalApi
    override fun toOffsetLimit(
        totalCount: Int,
        defaultPageSize: Int
    ): OffsetLimit =
        ConnectionArgumentsSupport.backwardOffsetLimit(last, before, totalCount, defaultPageSize).let {
            OffsetLimit(offset = it.offset, limit = it.limit)
        }

    /**
     * Validates backward pagination arguments.
     *
     * @throws IllegalArgumentException if last is not positive or before cursor is invalid
     */
    @ExperimentalApi
    override fun validate() = ConnectionArgumentsSupport.validateBackward(last, before)
}
